### PR TITLE
UIOAIPMH-12: Add set form validation

### PR DESCRIPTION
**Pull request by Dmitriy-Litvinenko**
# Purpose
Add set form validation
# Link
https://issues.folio.org/browse/UIOAIPMH-12
# Screenshot
image
image
image

### CODE REVIEW

**Comment by carolegodfrey** [comment]
Validations look good

One thing I did notice -- when you put in a lengthy set name (up to 255 characters) -- there is crowding on the view page such as below image -- separate but somewhat related issue

Screen Shot 2020-09-16 at 5 05 43 PM

File: test/bigtest/tests/sets/validation-test.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            ...initialValues,
+            filteringConditions: [],
+          });
+
+          this.visit(currentPath);
+
+          await triggerChangeAfterLoad();
+        });
+
+        it('should show success callout', () => {
```
[the remark this answers is not in the document]
Comment (Dmitriy-Litvinenko) [reply test/bigtest/tests/sets/validation-test.js]: Fixed

**Comment by Dmitriy-Litvinenko** [comment]
> Validations look good
> 
> One thing I did notice -- when you put in a lengthy set name (up to 255 characters) -- there is crowding on the view page such as below image -- separate but somewhat related issue
> 
> Screen Shot 2020-09-16 at 5 05 43 PM

**Comment by Dmitriy-Litvinenko** [comment]
> Validations look good
> 
> One thing I did notice -- when you put in a lengthy set name (up to 255 characters) -- there is crowding on the view page such as below image -- separate but somewhat related issue
> 
> Screen Shot 2020-09-16 at 5 05 43 PM

Fixed

**Comment by Dmitriy-Litvinenko** [comment]
@carolegodfrey  Could you please re-review?

File: src/settings/util/validation.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  EMPTY_VALIDATION_STATE_FOR_FIELD,
+} from '../constants';
+
+export const isFilteringConditionsFilled = (filteringConditions) => (
+  !!filteringConditions.filter((filteringCondition) => (
+    filteringCondition[FILTERING_CONDITIONS_FIELDS.NAME] &&
+    filteringCondition[FILTERING_CONDITIONS_FIELDS.ACTIVE] &&
+    filteringCondition[FILTERING_CONDITIONS_FIELDS.SET_SPEC] &&
+    filteringCondition[FILTERING_CONDITIONS_FIELDS.VALUE]
+  )).length
```
Comment (ViktorSoroka07) [inline src/settings/util/validation.js:17]:
```suggestion
  filteringConditions.some((filteringCondition) => (
    filteringCondition[FILTERING_CONDITIONS_FIELDS.NAME] &&
    filteringCondition[FILTERING_CONDITIONS_FIELDS.ACTIVE] &&
    filteringCondition[FILTERING_CONDITIONS_FIELDS.SET_SPEC] &&
    filteringCondition[FILTERING_CONDITIONS_FIELDS.VALUE]
  ))
```

File: src/settings/util/handleErrorResponse.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+      showCallout({
+        type: CALLOUT_ERROR_TYPE,
+        message: <FormattedMessage id={`${getTranslationKeyForError(type, field)}`} />,
+      });
+    });
+  } else {
+    showCallout({
+      type: CALLOUT_ERROR_TYPE,
+      message: <FormattedMessage id={`${UNSUPPORTED_ERROR_TRANSLATION_KEY}`} />,
```
Comment (ViktorSoroka07) [inline src/settings/util/handleErrorResponse.js:50]:
```suggestion
      message: <FormattedMessage id={UNSUPPORTED_ERROR_TRANSLATION_KEY} />,
```

File: src/settings/util/handleErrorResponse.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    const errorsResponse = await response.json();
+    const errors = get(errorsResponse, ['errors'], []);
+
+    errors.forEach((error) => {
+      const type = get(error, ['type'], '');
+      const field = get(error, ['parameters', 0, 'key'], '');
+
+      showCallout({
+        type: CALLOUT_ERROR_TYPE,
+        message: <FormattedMessage id={`${getTranslationKeyForError(type, field)}`} />,
```
Comment (ViktorSoroka07) [inline src/settings/util/handleErrorResponse.js:44]:
```suggestion
        message: <FormattedMessage id={getTranslationKeyForError(type, field)} />,
```

File: src/settings/util/validation.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        values={{
+          maxLength,
+        }}
+      />
+    );
+
+    return str.length <= maxLength ? EMPTY_VALIDATION_STATE_FOR_FIELD : errorMsg;
+  }
+
+  return EMPTY_VALIDATION_STATE_FOR_FIELD;
```
Comment (ViktorSoroka07) [inline src/settings/util/validation.js:39]: I would suggest dealing with error message stuff only when needed.

```suggestion
  const str = value.toString();

  if (value && str.length > maxLength) {
    return (
      <FormattedMessage
        id="ui-oai-pmh.settings.sets.edit.validationMessage.maxLength"
        values={{
          maxLength,
        }}
      />
    );
  }

  return EMPTY_VALIDATION_STATE_FOR_FIELD;
```

File: src/settings/components/Sets/SetsForm/components/GeneralInformation/GeneralInformation.js [hunk 6]
```diff
@@ -13,14 +15,27 @@ import {
...
 } from '../../../../../constants';
 
+import {
+  validateRequiredField,
+  validateFieldLength,
+  composeValidators,
+} from '../../../../../util';
+
+const validateNameLength = (value) => validateFieldLength(value, TEXT_FIELD_MAX_FIELD_LENGTH);
+const validateDescription = (value) => validateFieldLength(value, TEXT_AREA_FIELD_MAX_FIELD_LENGTH);
```
Comment (ViktorSoroka07) [inline src/settings/components/Sets/SetsForm/components/GeneralInformation/GeneralInformation.js:32]: For constistency with the name above.
```suggestion
const validateDescriptionLength = (value) => validateFieldLength(value, TEXT_AREA_FIELD_MAX_FIELD_LENGTH);
```

File: src/settings/components/Sets/SetsForm/components/FilteringConditions/FilteringConditions.js [hunk 7]
```diff
@@ -89,7 +124,9 @@ const FilteringConditions = ({
               <Field
                 id={`${FILTERING_CONDITIONS_FIELDS.ACTIVE}${index}`}
                 name={`${filteringCondition}.${FILTERING_CONDITIONS_FIELDS.ACTIVE}`}
+                aria-label="filteringConditionActive"
```
Comment (ViktorSoroka07) [inline src/settings/components/Sets/SetsForm/components/FilteringConditions/FilteringConditions.js:127]: Could you please clarify the reason you are adding aria-label for fields?

File: src/settings/util/handleErrorResponse.js (hunk 3 above)
Comment (Dmitriy-Litvinenko) [reply src/settings/util/handleErrorResponse.js:50]: Fixed

File: src/settings/util/handleErrorResponse.js (hunk 4 above)
Comment (Dmitriy-Litvinenko) [reply src/settings/util/handleErrorResponse.js:44]: Fixed

File: src/settings/components/Sets/SetsForm/components/GeneralInformation/GeneralInformation.js (hunk 6 above)
Comment (Dmitriy-Litvinenko) [reply src/settings/components/Sets/SetsForm/components/GeneralInformation/GeneralInformation.js:32]: Fixed

File: src/settings/components/Sets/SetsForm/components/FilteringConditions/FilteringConditions.js (hunk 7 above)
Comment (Dmitriy-Litvinenko) [reply src/settings/components/Sets/SetsForm/components/FilteringConditions/FilteringConditions.js:127]: For fix axe issue
image

Comment (ViktorSoroka07) [reply src/settings/components/Sets/SetsForm/components/FilteringConditions/FilteringConditions.js:127]: aria-labels should be readable for the users. The recommended approach, in this case, is to add further translations in the en.json file and then pass the localized result via intl.formatMessage.

File: src/settings/util/validation.js (hunk 2 above)
Comment (Dmitriy-Litvinenko) [reply src/settings/util/validation.js:17]: Fixed

File: src/settings/util/validation.js (hunk 5 above)
Comment (Dmitriy-Litvinenko) [reply src/settings/util/validation.js:39]: Fixed

File: src/settings/components/Sets/SetsForm/components/FilteringConditions/FilteringConditions.js (hunk 7 above)
Comment (Dmitriy-Litvinenko) [reply src/settings/components/Sets/SetsForm/components/FilteringConditions/FilteringConditions.js]: Fixed

**Comment by Dmitriy-Litvinenko** [comment]
@ViktorSoroka07 Could you please re-review?

**Review by ViktorSoroka07 — APPROVED** [review]
Thank you for handing PR comments @Dmitriy-Litvinenko.

**Review by carolegodfrey — APPROVED** [review]

**Pull request merged**